### PR TITLE
Switch from serializing Declarations to Serializing ComProjects.

### LINKED ISSUE #4390: Serialized declarations seem to be missing default member attributes.
Noticed this when I was writing the (missing) tests for #4327. Given a test setup that uses the Excel reference, this code...

```
            var interestingDeclarations =
                State.AllDeclarations.Where(item =>
                    item.AsTypeDeclaration != null
                    && ClassModuleDeclaration.HasDefaultMember(item.AsTypeDeclaration));
```
...returns no results.  The declarations collected from the typelib are fine.

### PR DESCRIPTION
**Pull request by comintern**
### Why do this?

Good question. The dependency graphs inside the `Declaration` objects were getting more and more complicated, and the `SerializableDeclaration`s were getting left behind, with some of the work of wiring them up being done in the `LibraryReferencedDeclarationsCollector` and some in the `Declaration` derived classes. Maintaining these changes in all of the different places is difficult, was at times ignored (by myself included), and was impacting the accuracy of RD's unit tests that used them.

So, that left 3 options:
1. Continue to painfully maintain the `SerializableDeclaration`.
2. Serialize the `Declaration`s directly.
3. Serialize the `ComProject`s that generate them.

Option 1 was discussed above. I rejected option 2 for two reasons - first, it would still leave the LRDC completely untestable in that it relies on loading libraries. Second, there are a hell of a lot more places in the `Declaration`s and derived classes that would need data contracts and member attributes than there are in the `ComReflection` namespace. From a practical standpoint, we can now start writing the years worth of missing unit tests for the `ReferencedDeclarationsCollector`.

### Worth mentioning:

- At the last minute, I decided to leave the 'Serialize' button in the debug UI. It should function the same as before - parse, hit the button, and then collect the serializations from the target folder.
- There's a new `IComProjectSerializationProvider` interface that abstracts serialization operations in case we ever decide to do a non-XML implementation (website maybe?). Note that it isn't wired up via IoC right now because, well, there's only one provider implementation.
- Given that the serialized and library loaded projects share the bulk of their code now, the the `ReferencedDeclarationsCollector` is back, now as a base class.
- There was a bug in how the default interfaces were being collected, which impacted the default member resolution of classes. That is fixed via this PR, and it uncovered a couple passing tests that should have been failing. Those already have issues opened for them, and they are referenced on the ignored tests.
- The only code changes that should require re-serialization now are changes in the COM collected objects.
- Other things that I probably forgot.

Oh yeah, this closes #4390. It should also provide more reliable tests for anything that revolves around default member resolution.

### CODE REVIEW

File: Rubberduck.Core/UI/Command/MenuItems/CommandBars/SerializeProjectsCommandMenuItem.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        public SerializeProjectsCommandMenuItem(SerializeDeclarationsCommand command) : base(command)
+        {
+        }
+
+        public override Func<string> Caption { get { return () => "Serialize"; } }
+        public override string Key => "SerializeProjects";
+    }
+
+    public class SerializeDeclarationsCommand : CommandBase
+    {
```
[the remark this answers is not in the document]
Comment (comintern) [reply Rubberduck.Core/UI/Command/MenuItems/CommandBars/SerializeProjectsCommandMenuItem.cs]: Thanks, thought I caught all of those.

File: Rubberduck.Core/UI/Command/MenuItems/CommandBars/SerializeProjectsCommandMenuItem.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            if (!Directory.Exists(_serializationProvider.Target))
+            {
+                Directory.CreateDirectory(_serializationProvider.Target);
+            }
+
+            var toSerialize = new Dictionary<Guid, ComProject>();
+
+            foreach (var project in _state.ProjectsProvider.Projects().Select(proj => proj.Project))
+            {
+                foreach (var reference in project.References.Select(lib => new ReferenceInfo(lib)))
```
[the remark this answers is not in the document]
Comment (comintern) [reply Rubberduck.Core/UI/Command/MenuItems/CommandBars/SerializeProjectsCommandMenuItem.cs]: Fixed.

File: Rubberduck.Core/UI/Command/MenuItems/CommandBars/SerializeProjectsCommandMenuItem.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+                {
+                    var library = _comLibraryProvider.LoadTypeLibrary(reference.FullPath);
```
[the remark this answers is not in the document]
Comment (comintern) [reply Rubberduck.Core/UI/Command/MenuItems/CommandBars/SerializeProjectsCommandMenuItem.cs]: I'm confused - it's the last parameter in the ctor.

[the remark this answers is not in the document]
Comment (MDoerner) [reply Rubberduck.Core/UI/Command/MenuItems/CommandBars/SerializeProjectsCommandMenuItem.cs]: Apparently, I was still sleeping when I wrote this.

**Pull request merged**